services/shhext/service.go

```diff
@@ -38,9 +42,16 @@ type EnvelopeEventsHandler interface {
 	MailServerRequestExpired(common.Hash)
 }
 
+// LoadedFilter is a record containing data about an added filter on whisper
+type LoadedFilter struct {
+	SymKeyID string
+	FilterID string
+}
+
 // Service is a service that provides some additional Whisper API.
 type Service struct {
 	w              *whisper.Whisper
+	publicAPI      *whisper.PublicWhisperAPI
```

Comment:
I don't really like this idea. PublicAPI can have a reference to a service because it uses it (the API is just a facade) but the other way around it seems odd.

If there is no easy way to remove it now, let's add a comment with `TODO` explaining that we need to get rid of this.

Reply:
It is a bit more complex, currently `SshExt` had a reference to whisper public api.

So it's ShhExtPublicAPI -> WhisperPublicAPI.

Effectively this means that any logic that uses the WhisperPublicAPI needs to go through the RPC public API (i.e. Post/GetNewFilterMessages), so logic becomes scattered around two files arbitrarly (service/and public api currently both have business logic), and currently Service is not encapsulated (publicapi uses private fields freely `api.service.*`) and logic is in both places.

This changes the behavior to be:

ShhExtPublicAPI -> ShhExtService -> WhisperPublicAPI , which basically is saying:
ShhExtPublicAPI offers a public api (method signatures), Service has all the business logic and knows about wrhisper, while the publicAPI does not (you could potentially keep the same public interface and change the transport layer without touching PublicAPI).

I am happy to change it, but I think we need better defined boundaries, what do you think?

Reply:
> So it's ShhExtPublicAPI -> WhisperPublicAPI.

That's fine in my opinion because `ShhExtPublicAPI` acts as a decorator on some `WhisperPublicAPI` methods and adds some more. It makes sense. But the scattered logic is bad. Not sure why it ended up in this way.

> ShhExtPublicAPI -> ShhExtService -> WhisperPublicAPI , which basically is saying:
ShhExtPublicAPI offers a public api (method signatures), Service has all the business logic and knows about wrhisper, while the publicAPI does not (you could potentially keep the same public interface and change the transport layer without touching PublicAPI).

Everything about `ShhExtPublicAPI` makes sense and if it could only use public methods of `ShhExtService` that would be even better (better encapsulation). Finally, `ShhExtService` could be a programmable interface and `ShhExtPublicAPI` only as a facade needed by the RPC server.

What I would like to change is to have `ShhExtPublicAPI -> ShhExtService -> WhisperService`. Unless `WhisperService` is not really well implemented and a lot of logic is put inside `WhisperPublicAPI`... If this is the case, let's just put a comment with explanation. If not, we can create an issue and tackle it in a different PR (I would not do it in this one).